File: app/src/main/java/com/ferid/app/frequentcontacts/selectnumber/SelectNumberActivity.java
/*
 * Copyright (C) 2016 Ferid Cafer
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ferid.app.frequentcontacts.selectnumber;

import android.app.SearchManager;
import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.provider.ContactsContract;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.util.Base64;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.ferid.app.frequentcontacts.R;
import com.ferid.app.frequentcontacts.list.Contact;
import com.pnikosis.materialishprogress.ProgressWheel;

import java.text.Collator;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Locale;


/**
 * Created by ferid.cafer on 11/12/2014.
 */
public class SelectNumberActivity extends AppCompatActivity {

    private ArrayList<Contact> numberList = new ArrayList<>();
    private NumberAdapter adapter;

    private ProgressWheel progressWheel;

    private ArrayList<Contact> wholeArrayList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.select_number);

        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);

        ListView list = (ListView) findViewById(R.id.list);
        adapter = new NumberAdapter(this, R.layout.select_number_row, numberList);

        list.setAdapter(adapter);

        list.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int position, long l) {
                Contact contact = numberList.get(position);

                //retrieve and set photo of the selected contact
                contact.setPhoto(retrievePhoto(contact.getId()));

                Intent intent = new Intent();
                intent.putExtra("contact", contact);
                setResult(RESULT_OK, intent);
                closeWindow();
            }
        });

        progressWheel = (ProgressWheel) findViewById(R.id.progressWheel);


        new NumberListRetriever().execute();
    }

    /**
     * Get names and number of contacts
     * @return
     */
    private ArrayList<Contact> getNumbersList() {
        ArrayList<Contact> tmpList = new ArrayList<Contact>();
        Contact contact;

        ContentResolver cr = getContentResolver();
        Cursor cur = cr.query(ContactsContract.Contacts.CONTENT_URI, null, null, null, null);

        if (cur != null) {
            try {
                if (cur.getCount() > 0) {
                    while (cur.moveToNext()) {
                        //names
                        int id = cur.getInt(cur.getColumnIndex(ContactsContract.Contacts._ID));
                        final String name = cur.getString(
                                cur.getColumnIndex(ContactsContract.Contacts.DISPLAY_NAME));

                        if (Integer.parseInt(
                                cur.getString(cur.getColumnIndex(
                                        ContactsContract.Contacts.HAS_PHONE_NUMBER))) > 0) {
                            //phones
                            Cursor phones = getContentResolver().query(
                                    ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                                    new String[]{ContactsContract.CommonDataKinds.Phone.NUMBER,
                                            ContactsContract.CommonDataKinds.Phone.TYPE},
                                    ContactsContract.CommonDataKinds.Phone.CONTACT_ID + " = " + id,
                                    null, null);

                            if (phones != null) {
                                while (phones.moveToNext()) {
                                    String phoneNumber = phones.getString(
                                            phones.getColumnIndex(
                                                    ContactsContract.CommonDataKinds.Phone.NUMBER));

                                    //create a new Contact object
                                    contact = new Contact();
                                    contact.setId(id);
                                    contact.setName(name);
                                    contact.setNumber(phoneNumber);

                                    if (!tmpList.contains(contact)) {
                                        tmpList.add(contact);
                                    }
                                }
                                phones.close();
                            }
                        }
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                cur.close();
            }
        }

        return tmpList;
    }

    /**
     * Retrieve photo of the given contact with contact ID
     * @param contactId
     * @return
     */
    private String retrievePhoto(int contactId) {
        Uri contactUri = ContentUris.withAppendedId(ContactsContract.Contacts.CONTENT_URI,
                contactId);
        Uri photoUri = Uri.withAppendedPath(contactUri,
                ContactsContract.Contacts.Photo.CONTENT_DIRECTORY);

        Cursor cursor = getContentResolver().query(photoUri,
                new String[]{ContactsContract.Contacts.Photo.PHOTO}, null, null, null);

        if (cursor == null) {
            return "";
        }

        try {
            if (cursor.moveToFirst()) {
                byte[] data = cursor.getBlob(0);
                if (data != null) {
                    //byte[] to String convertion
                    String photo = Base64.encodeToString(data, Base64.DEFAULT);
                    return photo;
                }
            }
        } finally {
            cursor.close();
        }

        return "";
    }

    /**
     * Read data and refresh the list
     */
    private class NumberListRetriever extends AsyncTask<Void, Void, ArrayList<Contact>> {

        @Override
        protected void onPreExecute() {
            progressWheel.spin();
        }

        @Override
        protected ArrayList<Contact> doInBackground(Void... params) {
            //retrieve elements
            ArrayList<Contact> tmpList = getNumbersList();

            //sort elements
            if (tmpList != null) {
                Collections.sort(tmpList, new ContactsComparator());
            }

            //send to UI thread
            return tmpList;
        }

        @Override
        protected void onPostExecute(ArrayList<Contact> result) {
            if (result != null) {
                numberList.addAll(result);
                wholeArrayList.addAll(result);
                adapter.notifyDataSetChanged();
            }

            if (progressWheel.isSpinning()) progressWheel.stopSpinning();
        }
    }

    /**
     * Alphabetical order
     */
    public class ContactsComparator implements Comparator<Contact> {
        //prepare utf-8
        Collator collator = Collator.getInstance(new Locale("UTF-8"));

        @Override
        public int compare(Contact c1, Contact c2) {
            return collator.compare(c1.getName(), c2.getName());
        }
    }

    /**
     * Run time searching
     * @param searchText
     */
    private void searchEngine(final String searchText) {
        Handler handler = new Handler();
        handler.post(new Runnable() {
            @Override
            public void run() {
                ArrayList<Contact> tmpList = new ArrayList<Contact>();

                //first, add names that start with searchText
                for (Contact location : wholeArrayList) {
                    if (location.getName().toLowerCase().startsWith(searchText.toLowerCase())) {
                        tmpList.add(location);
                    }
                }
                //then, add names that contain searchText
                for (Contact location : wholeArrayList) {
                    if (location.getName().toLowerCase().contains(searchText.toLowerCase())
                            && !location.getName().toLowerCase().startsWith(searchText.toLowerCase())) {
                        tmpList.add(location);
                    }
                }

                numberList.clear();
                numberList.addAll(tmpList);
                adapter.notifyDataSetChanged();
            }
        });
    }

    private void closeWindow() {
        finish();
        overridePendingTransition(R.anim.stand_still, R.anim.move_out_to_bottom);
    }

    @Override
    public void onBackPressed() {
        closeWindow();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search, menu);

        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        SearchView searchView = (SearchView) menu.findItem(R.id.ic_action_search).getActionView();
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                return true;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                if (s.length() >= 1)
                    searchEngine(s);
                else {
                    numberList.clear();
                    numberList.addAll(wholeArrayList);
                    adapter.notifyDataSetChanged();
                }
                return true;
            }
        });

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar actions click
        switch (item.getItemId()) {
            case android.R.id.home:
                closeWindow();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/com/ferid/app/frequentcontacts/selectnumber/SelectNumberActivity.kt
/*
 * Copyright (C) 2014 Ferid Cafer
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ferid.app.frequentcontacts.selectnumber

import android.Manifest
import android.app.Activity
import android.app.SearchManager
import android.content.ContentResolver
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.database.Cursor
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.provider.ContactsContract
import android.text.TextUtils
import android.util.Base64
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.ferid.app.frequentcontacts.R
import com.ferid.app.frequentcontacts.list.Contact
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.activity_select_number.toolbar
import kotlinx.android.synthetic.main.activity_select_number.*
import java.text.Collator
import java.util.*
import kotlin.collections.ArrayList

class SelectNumberActivity : AppCompatActivity() {

    private var progressBar: ProgressBar? = null

    private lateinit var mContext: Context
    private lateinit var mAdapter: NumberAdapter
    private var mContactList: ArrayList<Contact> = ArrayList()

    //whole number list, used for searching any contact
    private val wholeArrayList = java.util.ArrayList<Contact>()

    private val requestPermissions: Int = 100

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_select_number)

        mContext = this

        setSupportActionBar(toolbar as Toolbar?)

        if (actionBar != null) {
            actionBar!!.setDisplayHomeAsUpEnabled(true)
            actionBar!!.setHomeButtonEnabled(true)
        }

        progressBar = findViewById(R.id.progressBar)

        initRecyclerView()

        if (ContextCompat.checkSelfPermission(this,
                        Manifest.permission.READ_CONTACTS)
                == PackageManager.PERMISSION_GRANTED &&
                ContextCompat.checkSelfPermission(this,
                        Manifest.permission.CALL_PHONE)
                == PackageManager.PERMISSION_GRANTED) { //if permission is already granted

            getContactsList()
        } else {
            getPermissions()
        }
    }

    /**
     * Initialise recycler view
     */
    private fun initRecyclerView() {
        rvList.layoutManager = LinearLayoutManager(mContext)
        mAdapter = NumberAdapter(mContactList) {
            setItemSelection(it)
        }
        rvList.adapter = mAdapter
    }

    /**
     * Set selected contact. Retrieve photo if exists and close window.
     * @param contact Selected contact
     */
    private fun setItemSelection(contact: Contact) {
        //retrieve and set photo of the selected contact
        contact.photo = retrievePhoto(contact.id)

        val intent = Intent()
        intent.putExtra("contact", contact)
        setResult(Activity.RESULT_OK, intent)
        closeWindow()
    }

    /**
     * Retrieve list of contacts
     */
    private fun getContactsList() {
        NumberListRetriever(mContext, progressBar!!, mAdapter, mContactList, wholeArrayList).execute()
    }

    /**
     * Request and get the permission for reading contacts and phone calling
     */
    private fun getPermissions() {
        if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                        Manifest.permission.READ_CONTACTS)
                || ActivityCompat.shouldShowRequestPermissionRationale(this,
                        Manifest.permission.CALL_PHONE)) {
            Snackbar.make(progressBar!!, R.string.grantPermission,
                    Snackbar.LENGTH_LONG)
                    .setAction(R.string.ok) {
                        ActivityCompat.requestPermissions(this@SelectNumberActivity,
                                arrayOf(Manifest.permission.READ_CONTACTS,
                                Manifest.permission.CALL_PHONE),
                                requestPermissions)
                    }.show()
        } else {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_CONTACTS,
                    Manifest.permission.CALL_PHONE),
                    requestPermissions)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == requestPermissions) {
            if (grantResults.size > 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED
                    && grantResults[1] == PackageManager.PERMISSION_GRANTED) {

                //permission granted
                getContactsList()
            } else {
                //permission denied
                closeWindow()
            }
        }
    }

    /**
     * Retrieve photo of the given contact with contact ID
     * @param contactId
     * @return
     */
    private fun retrievePhoto(contactId: Int): String? {
        val contactUri = ContentUris.withAppendedId(ContactsContract.Contacts.CONTENT_URI,
                contactId.toLong())
        val photoUri = Uri.withAppendedPath(contactUri,
                ContactsContract.Contacts.Photo.CONTENT_DIRECTORY)
        val cursor: Cursor = contentResolver.query(photoUri,
                arrayOf(ContactsContract.Contacts.Photo.PHOTO),
                null, null, null)
                ?: return ""
        cursor.use { cursor ->
            if (cursor.moveToFirst()) {
                val data = cursor.getBlob(0)
                if (data != null) {
                    //byte[] to String conversion
                    return Base64.encodeToString(data, Base64.DEFAULT)
                }
            }
        }
        return ""
    }

    /**
     * Read data and refresh the list
     */
    class NumberListRetriever(var context: Context, var progressBar: ProgressBar, var adapter: NumberAdapter,
                              var numberList: ArrayList<Contact>, var wholeArrayList: ArrayList<Contact>)
        : AsyncTask<Void?, Void?, ArrayList<Contact>?>() {

        /**
         * Get names and number of contacts
         * @return
         */
        private fun getNumbersList(): ArrayList<Contact> {
            //keep track of added phones in order to handle redundancy
            val addedPhoneNumbers = ArrayList<String?>()

            //contacts to be showed
            val tmpList: ArrayList<Contact> = ArrayList()
            var contact: Contact
            val cr: ContentResolver = context.contentResolver
            val cur = cr.query(ContactsContract.Contacts.CONTENT_URI, null, null, null, null)
            cur?.use { cur ->
                if (cur.count > 0) {
                    while (cur.moveToNext()) {
                        //names
                        val id = cur.getInt(cur.getColumnIndex(ContactsContract.Contacts._ID))
                        val name = cur.getString(
                                cur.getColumnIndex(ContactsContract.Contacts.DISPLAY_NAME))
                        if (
                                cur.getString(cur.getColumnIndex(
                                        ContactsContract.Contacts.HAS_PHONE_NUMBER)).toInt() > 0) {
                            //phones
                            val phones = context.contentResolver.query(
                                    ContactsContract.CommonDataKinds.Phone.CONTENT_URI, arrayOf(ContactsContract.CommonDataKinds.Phone.NUMBER,
                                    ContactsContract.CommonDataKinds.Phone.TYPE),
                                    ContactsContract.CommonDataKinds.Phone.CONTACT_ID + " = " + id,
                                    null, null)
                            if (phones != null) {
                                while (phones.moveToNext()) {
                                    val phoneNumber = phones.getString(
                                            phones.getColumnIndex(
                                                    ContactsContract.CommonDataKinds.Phone.NUMBER))

                                    //create a new Contact object
                                    contact = Contact()
                                    contact.id = id
                                    contact.name = name
                                    contact.number = phoneNumber.replace(" ", "")
                                    if (!TextUtils.isEmpty(contact.number)
                                            && !addedPhoneNumbers.contains(contact.number)) {
                                        tmpList.add(contact)
                                        addedPhoneNumbers.add(contact.number)
                                    }
                                }
                                phones.close()
                            }
                        }
                    }
                }
            }

            return tmpList
        }

        override fun onPreExecute() {
            progressBar.visibility = View.VISIBLE
        }

        override fun doInBackground(vararg p0: Void?): ArrayList<Contact>? {
            //retrieve elements
            val tmpList: ArrayList<Contact> = getNumbersList()

            //sort elements
            Collections.sort(tmpList, ContactsComparator())
            val contact = Contact()
            contact.name = "None"
            tmpList.add(0, contact)

            //send to UI thread
            return tmpList
        }

        override fun onPostExecute(result: ArrayList<Contact>?) {
            if (result != null) {
                numberList.addAll(result)
                wholeArrayList.addAll(result)
                adapter.notifyDataSetChanged()
            }
            progressBar.setVisibility(View.GONE)
        }
    }

    /**
     * Alphabetical order
     */
    class ContactsComparator : Comparator<Contact?> {
        //prepare utf-8
        var collator = Collator.getInstance(Locale("UTF-8"))!!

        override fun compare(c1: Contact?, c2: Contact?): Int {
            return collator.compare(c1!!.name, c2!!.name)
        }
    }

    /**
     * Run time searching
     * @param searchText
     */
    private fun searchEngine(searchText: String) {
        val handler = Handler()
        handler.post {
            val tmpList = java.util.ArrayList<Contact>()

            //first, add names that start with searchText
            for (location in wholeArrayList) {
                if (location.name!!.toLowerCase(Locale.getDefault()).startsWith(
                                searchText.toLowerCase(Locale.getDefault()))) {
                    tmpList.add(location)
                }
            }
            //then, add names that contain searchText
            for (location in wholeArrayList) {
                if (location.name!!.toLowerCase(Locale.getDefault()).contains(
                                searchText.toLowerCase(Locale.getDefault()))
                        && !location.name!!.toLowerCase(Locale.getDefault()).startsWith(
                                searchText.toLowerCase(Locale.getDefault()))) {
                    tmpList.add(location)
                }
            }
            mContactList.clear()
            mContactList.addAll(tmpList)
            mAdapter.notifyDataSetChanged()
        }
    }

    private fun closeWindow() {
        finish()
        overridePendingTransition(R.anim.stand_still, R.anim.move_out_to_bottom)
    }

    override fun onBackPressed() {
        closeWindow()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.ic_action_search).actionView as SearchView
        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                return true
            }

            override fun onQueryTextChange(s: String): Boolean {
                if (s.isNotEmpty()) searchEngine(s) else {
                    mContactList.clear()
                    mContactList.addAll(wholeArrayList)
                    mAdapter.notifyDataSetChanged()
                }
                return true
            }
        })
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            closeWindow()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}